subsys/usb/usb_descriptor.c

```diff
@@ -290,16 +291,30 @@ static struct usb_cfg_data *usb_get_cfg_data(struct usb_if_descriptor *iface)
 }
 
 /*
- * Default USB SN string descriptor is CONFIG_USB_DEVICE_SN, but sometimes
- * we want use another string as SN descriptor such as the chip's unique
- * ID. So user can implement this function return a string thats will be
- * replaced the default SN.
- * Please note that the new SN descriptor you changed must has same length
- * as CONFIG_USB_DEVICE_SN.
+ * Default USB Serial Number string descriptor will be derived from
+ * Hardware Information Driver (HWINFO). User can implement own variant
+ * of this function. Please note that the length of the new Serial Number
+ * descriptor may not exceed the length of the CONFIG_USB_DEVICE_SN.
  */
 __weak u8_t *usb_update_sn_string_descriptor(void)
 {
-	return NULL;
+	u8_t hwid[sizeof(CONFIG_USB_DEVICE_SN) / 2];
```

Comment:
You should initialize this variable. The hwinfo_get_device_id might not fill the whole array.
You could also check the return value to match the size of the hwid[].
Also, mind that the zero termination is included in the `sizeof(CONFIG_USB_DEVICE_SN)`.

Reply:
> You should initialize this variable. The hwinfo_get_device_id might not fill the whole array.

L:306

> You could also check the return value to match the size of the hwid[].

L:309

Reply:
`hwinfo_get_device_id` returns a positive number with the length. But you are right, there is nothing you can do about if the number is too short.

Reply:
sn buffer is too long, will fix

Reply:
Actually it does a implicit round up.
u8_t hwid[((size of the string) + 1) / 2];